src/contracts/atlas/GasAccounting.sol

```diff
@@ -70,21 +80,34 @@ abstract contract GasAccounting is SafetyLocks {
         if (maxApprovedGasSpend > bondedBalance) maxApprovedGasSpend = bondedBalance;
 
         if (lock != environment) revert InvalidExecutionEnvironment(lock);
-        if (solverFrom != solver) revert InvalidSolverFrom(solver);
 
-        uint256 _deposits = deposits + msg.value;
+        (address currentSolver, bool verified, bool fulfilled) = solverLockData();
+
+        if (solverFrom != currentSolver) revert InvalidSolverFrom(currentSolver);
 
         uint256 deficit = claims + withdrawals;
-        uint256 surplus = _deposits + maxApprovedGasSpend;
+        uint256 surplus = deposits + maxApprovedGasSpend + msg.value;
 
-        if (deficit > surplus) revert InsufficientTotalBalance(deficit - surplus);
+        // if (deficit > surplus) revert InsufficientTotalBalance(deficit - surplus);
 
         // Add msg.value to solver's deposits
-        if (msg.value > 0) deposits = _deposits;
+        if (msg.value > 0 || maxApprovedGasSpend > 0) deposits = surplus;
 
-        solver = SOLVER_FULFILLED;
+        // CASE: Callback verified but insufficient balance
+        if (deficit > surplus) {
+            if (!verified) {
+                uint256 solverLock = uint256(uint160(currentSolver)) | _solverVerified;
+                _solverLock = solverLock;
```

Comment:
```suggestion
                _solverLock = uint256(uint160(currentSolver)) | _solverVerified;
```
Same gas efficiency thing here - looks like it improves the gas cost in the snapshot of `test_reconcile`